File: sensor_logger.go
// This code listen port 1703 and get data from ESP-32 in "#RTNode1=1111" format

package main

import (
    "fmt"
    "net"
    "os"
    "strings"
    "time"
)

const (
    UDP_IP      = "0.0.0.0"
    UDP_PORT    = 1703
    BUFFER_SIZE = 32
    LOG_FILE    = "sensor_data.log"
)

func logData(data string) {
    file, err := os.OpenFile(LOG_FILE, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
    if err != nil {
        fmt.Println("Error opening log file:", err)
        return
    }
    defer file.Close()

    timestamp := time.Now().Format("2006-01-02 15:04:05")
    _, err = file.WriteString(fmt.Sprintf("%s - %s\n", timestamp, data))
    if err != nil {
        fmt.Println("Error writing to log file:", err)
    }
}

func parseString(inputStr string) []string {
    inputStr = strings.TrimPrefix(inputStr, "#")
    parts := strings.Split(inputStr, "=")
    if len(parts) != 2 {
        return nil
    }

    nodePart := parts[0]
    valuesPart := strings.Split(parts[1], "")
    result := append([]string{nodePart}, valuesPart...)

    return result
}

func main() {
    addr := net.UDPAddr{
        Port: UDP_PORT,
        IP:   net.ParseIP(UDP_IP),
    }

    conn, err := net.ListenUDP("udp", &addr)
    if err != nil {
        fmt.Println("Error creating UDP socket:", err)
        return
    }
    defer conn.Close()

    fmt.Printf("Listening on %s:%d...\n", UDP_IP, UDP_PORT)

    buffer := make([]byte, BUFFER_SIZE)

    for {
        n, clientAddr, err := conn.ReadFromUDP(buffer)
        if err != nil {
            fmt.Println("Error reading from UDP:", err)
            continue
        }

        dataStr := strings.TrimSpace(string(buffer[:n]))

        logData(dataStr)

        parsedList := parseString(dataStr)

        fmt.Printf("Received message from %s: %s\n", clientAddr, dataStr)
        fmt.Printf("Parsed message: %v\n", parsedList)

		for i := 1; i < len(parsedList); i++ {
			fmt.Printf("Received message, item %d: %s\n", i, parsedList[i])
		}
    }
}

File: sensor_logger.py
# This code listen port 1703 and get data from ESP-32 in "#RTNode1=1111" format

import socket
import datetime

# Server parameters
UDP_IP = "0.0.0.0"  # all interfaces
UDP_PORT = 1703
BUFFER_SIZE = 32  # buffer size

# Log file name
LOG_FILE = "sensor_data.log"

# Write on file
def log_data(data):
    with open(LOG_FILE, "a") as file:
        timestamp = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        file.write(f"{timestamp} - {data}\n")

# Create UDP socket
sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
sock.bind((UDP_IP, UDP_PORT))

print(f"Listening on {UDP_IP}:{UDP_PORT}...")

def parse_string(input_str):
    # Delete '#' and split by '='
    node_part, values_part = input_str[1:].split('=')
    
    # Convert value into list
    values_list = list(values_part)
    
    # Combine the part with the node name and the list of values
    result = [node_part] + values_list
    
    return result

try:
    while True:
        # Get the data from client (ESP32)
        data, addr = sock.recvfrom(BUFFER_SIZE)
        data_str = data.decode('utf-8').strip()
        
        # logging the data
        log_data(data_str)

        parsed_list = parse_string(data_str)

        print(f"Received message from {addr}: {data_str}")
        print(f"Parsed message from {addr}: {parsed_list}")        

except KeyboardInterrupt:
    print("Server stopped.")

finally:
    # Close socket in the end
    sock.close()
